src/H5Olayout.c

```diff
@@ -634,13 +634,22 @@ H5O__layout_decode(H5F_t *f, H5O_t H5_ATTR_UNUSED *open_oh, unsigned H5_ATTR_UNU
                         heap_block_p += tmp_size;
 
                         /* Source selection */
-                        if (H5S_SELECT_DESERIALIZE(&mesg->storage.u.virt.list[i].source_select,
-                                                   &heap_block_p) < 0)
+                        if (heap_block_p_end - heap_block_p + 1 <= 0)
```

Comment:
Suggest doing the math here once and storing it in a `ptrdiff_t` variable just for future maintenance reasons.